src/util/simplify_expr_if.cpp

```diff
@@ -260,21 +279,18 @@ bool simplify_exprt::simplify_if_preorder(if_exprt &expr)
       }
     }
     else
-      local_replace_map.insert(std::make_pair(cond, true_exprt()));
+      local_replace_map.insert(std::make_pair(r_cond.expr, true_exprt()));
 
-    auto r_truevalue = simplify_rec(truevalue);
-    if(r_truevalue.has_changed())
-    {
-      truevalue = r_truevalue.expr;
-      no_change = false;
-    }
+    auto r_truevalue = simplify_rec(swap_branches ? falsevalue : truevalue);
+    if(swap_branches)
+      r_truevalue.expr_changed = CHANGED;
 
```

Comment:
Why is this needed? The if-expression will be marked as changed anyway.

Reply:
Yes, but `build_if_expr` would not actually replace the `truevalue`.

Reply:
I have refactored the code and added a comment.